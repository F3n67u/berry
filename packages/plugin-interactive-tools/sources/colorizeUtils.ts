import {Configuration, formatUtils, structUtils} from '@yarnpkg/core';
import {diffWords}                               from 'diff';

const SIMPLE_SEMVER = /^((?:[\^~]|>=?)?)([0-9]+)(\.[0-9]+)(\.[0-9]+)((?:-\S+)?)$/;

export function colorizeVersionDiff(configuration: Configuration, from: string, to: string) {
  if (from === to)
    return to;

  const parsedFrom = structUtils.parseRange(from);
  const parsedTo = structUtils.parseRange(to);

  const matchedFrom = parsedFrom.selector.match(SIMPLE_SEMVER);
  const matchedTo = parsedTo.selector.match(SIMPLE_SEMVER);

  if (!matchedFrom || !matchedTo)
    return colorizeRawDiff(configuration, from, to);

  const SEMVER_COLORS = [
    `gray`,
    `red`,
    `yellow`,
    `green`,
    `magenta`, // prerelease
  ];

  let color: string | null = null;
  let res = ``;

  for (let t = 1; t <= SEMVER_COLORS.length; ++t) {
    if (color !== null || matchedFrom[t] !== matchedTo[t]) {
      if (color === null)
        color = SEMVER_COLORS[t - 1];

      res += formatUtils.pretty(configuration, matchedTo[t], color);
    } else {
      res += matchedTo[t];
    }
  }

  return res;
}

function colorizeRawDiff(configuration: Configuration, from: string, to: string) {
  const diff = diffWords(from, to);
  let str = ``;

  for (const part of diff) {
    if (part.added) {
      str += formatUtils.pretty(configuration, part.value, `green`);
    } else if (!part.removed) {
      str += part.value;
    }
  }

  return str;
}
