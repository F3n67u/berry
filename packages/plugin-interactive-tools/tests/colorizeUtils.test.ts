import {Configuration}       from '@yarnpkg/core';
import {xfs, PortablePath}   from '@yarnpkg/fslib';

import {colorizeVersionDiff} from '../sources/colorizeUtils';

async function initializeConfiguration<T>(value: {[key: string]: any}, cb: (dir: PortablePath) => Promise<T>) {
  return await xfs.mktempPromise(async dir => {
    await Configuration.updateConfiguration(dir, value);

    return await cb(dir);
  });
}

describe(`colorizeUtils`, () => {
  describe(`colorizeVersionDiff`, () => {
    it(`should keep the same version if there is no difference`, async () => {
      await initializeConfiguration({}, async dir => {
        const configuration = await Configuration.find(dir, null);
        expect(colorizeVersionDiff(configuration, `^2.0.0`, `^2.0.0`)).toBe(`^2.0.0`);
      });
    });

    it(`should handle semver without modifier part`, async () => {
      await initializeConfiguration({}, async dir => {
        const configuration = await Configuration.find(dir, null);
        expect(colorizeVersionDiff(configuration, `2.0.0`, `2.1.0`)).toMatchSnapshot();
      });
    });

    it(`should handle semver with prerelease part`, async () => {
      await initializeConfiguration({}, async dir => {
        const configuration = await Configuration.find(dir, null);
        expect(colorizeVersionDiff(configuration, `^2.0.0-beta.17`, `^2.0.0-beta.18`)).toMatchSnapshot();
      });
    });
  });
});
