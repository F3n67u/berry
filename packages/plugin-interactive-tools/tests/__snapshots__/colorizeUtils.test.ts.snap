// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorizeUtils colorizeVersionDiff should handle semver with prerelease part 1`] = `"^2.0.0[35m-beta.18[39m"`;

exports[`colorizeUtils colorizeVersionDiff should handle semver without modifier part 1`] = `"2[33m.1[39m[33m.0[39m"`;
